storage/range.go

```diff
@@ -452,32 +482,43 @@ func (r *Range) addReadOnlyCmd(method string, args proto.Request, reply proto.Re
 
 	// If read-consistency is set to INCONSISTENT, run directly.
 	if header.ReadConsistency == proto.INCONSISTENT {
+		// But disallow any inconsistent reads within txns.
+		if args.Header().ReadConsistency == proto.INCONSISTENT && args.Header().Txn != nil {
+			reply.Header().SetGoError(util.Errorf("cannot allow inconsistent reads within a transaction"))
+			return reply.Header().GoError()
+		}
 		return r.executeCmd(0, method, args, reply)
+	} else if args.Header().ReadConsistency == proto.CONSENSUS {
+		reply.Header().SetGoError(util.Errorf("consensus reads not implemented"))
+		return reply.Header().GoError()
+	}
+
+	held, prevTerm := r.HasLeaderLease()
+	if !held {
+		// If we don't hold the lease, but nobody else does either, submit
+		// this read-only command as read-write. The read will only succeed
+		// if this replica is the leader, and the command queue and read
+		// timestamp cache are both updated the same as if we had the lease.
+		if anotherHeld, _ := r.AnotherHasLeaderLease(); !anotherHeld {
```

Comment:
I know this part is going to change, but it looks like `AnotherHasLeaderLease`, if returning `anotherHeld == true`, should guarantee that that other node still has it, i.e. it should use the clock offset against and not for it. We lose a little bit there but if a lease is that close to expiring, then there's little point in recommending that leader to anyone.